types/mempool/mempool.go

```diff
@@ -0,0 +1,254 @@
+
+	// Remove attempts to remove a transaction from the mempool, returning an error
+	// upon failure.
+	Remove(Tx) error
+}
+
+type ErrTxNotFound struct {
+	error
+}
+
+type Factory func() Mempool
+
+var _ Mempool = (*defaultMempool)(nil)
+
+// defaultMempool is the SDK's default mempool implementation which stores txs in a partially ordered set
+// by 2 dimensions; priority, and sender-nonce.  Internally it uses one priority ordered skip list and one skip list
+// per sender ordered by nonce (sender-nonce).  When there are multiple txs from the same sender, they are not
+// always comparable by priority to other sender txs and must be partially ordered by both sender-nonce and priority.
+type defaultMempool struct {
+	priorities *huandu.SkipList
+	senders    map[string]*huandu.SkipList
+	scores     map[txKey]int64
+	iterations int
+}
+
+type txKey struct {
+	nonce    uint64
+	priority int64
+	sender   string
+}
+
+// txKeyLess is a comparator for txKeys that first compares priority, then sender, then nonce, uniquely identifying
+// a transaction.
+func txKeyLess(a, b interface{}) int {
+	keyA := a.(txKey)
+	keyB := b.(txKey)
+	res := huandu.Int64.Compare(keyA.priority, keyB.priority)
+	if res != 0 {
+		return res
+	}
+
+	res = huandu.String.Compare(keyA.sender, keyB.sender)
+	if res != 0 {
+		return res
+	}
+
+	return huandu.Uint64.Compare(keyA.nonce, keyB.nonce)
+}
+
+// NewDefaultMempool returns the SDK's default mempool implementation which returns txs in a partial order
+// by 2 dimensions; priority, and sender-nonce.
+func NewDefaultMempool() Mempool {
+	return &defaultMempool{
+		priorities: huandu.New(huandu.LessThanFunc(txKeyLess)),
+		senders:    make(map[string]*huandu.SkipList),
+		scores:     make(map[txKey]int64),
+	}
+}
+
+// Insert attempts to insert a Tx into the app-side mempool in O(log n) time, returning an error if unsuccessful.
+// Sender and nonce are derived from the transaction's first signature.
+// Transactions are unique by sender and nonce.
+// Inserting a duplicate tx is an O(log n) no-op.
+// Inserting a duplicate tx with a different priority overwrites the existing tx, changing the total order of
+// the mempool.
+func (mp *defaultMempool) Insert(ctx types.Context, tx Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	tk := txKey{nonce: nonce, priority: ctx.Priority(), sender: sender}
+
+	senderTxs, ok := mp.senders[sender]
+	// initialize sender mempool if not found
+	if !ok {
+		senderTxs = huandu.New(huandu.LessThanFunc(func(a, b interface{}) int {
+			return huandu.Uint64.Compare(b.(txKey).nonce, a.(txKey).nonce)
+		}))
+		mp.senders[sender] = senderTxs
+	}
+
+	// Since senderTxs is scored by nonce, a changed priority will overwrite the existing txKey.
+	senderTxs.Set(tk, tx)
```

Comment:
I don't understand this comment and use of `txKey`. From what I understand, the `senderIndex` concerns itself only with the nonce/sequence, right? Why do we need to consider or have priority be involved?

Reply:
It's an implementation detail.  Priority is carried in the key as metadata since the algorithm checks the priority of each tx in the sender index before selecting it to verify if the edge traversal is valid.  If it's not a higher priority tx from a different sender will be selected first.

It saves a (constant time) lookup to `mp.scores` and moreover seems good to use the same tx key in all indices.